Add threadpool to join the recovery threads

claddyy opened the pull request:
Fixes #284

Shourya742 commented:
Any specific reason for adding another crate for thread handling?

claddyy commented:
> Any specific reason for adding another crate for thread handling?

It was not needed, I tried implementing the methods from `threadpool` crate earlier, which was not needed. Refined the implementation now.

claddyy commented on src/maker/api.rs:
```diff
@@ -92,6 +91,42 @@ pub struct ConnectionState {
+}
+
+impl ThreadPool {
+    pub fn new() -> Self {
```
Addressed in f27d26e7083dc0bd87d13dc5d2d30a833e4eb239
This is better, than to manually join threads. Thank you!

claddyy commented on src/maker/api.rs:
```diff
@@ -92,6 +91,48 @@ pub struct ConnectionState {
+        threads.push(handle);
+    }
+
+    pub fn join_all(&self) -> Result<(), MakerError> {
```
Addressed [1ca49c6](https://github.com/citadel-tech/coinswap/pull/316/commits/1ca49c6cc61d7373719ef74c00c006ac854c619d)

KnowWhoami reviewed (changes requested):
PartialACK.
Here are the list of review  comments focusing on handling the threads and related errors in much better way.

KnowWhoami commented:
@Shourya742 
- we can't push the `tor_handle` inside the maker's thread_pool because tor_handle is of type `mitosis::JoinHandle` while thread_pool must contain `JoinHandle` of type `std::thread::Joinhandle`.
Though right now , this is not an issue but may become a potential bug when implementing `drop` trait for `Maker` struct.

KnowWhoami commented on src/maker/api.rs:
```diff
@@ -92,6 +91,48 @@ pub struct ConnectionState {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```
Do you mean to say that we also require Drop trait to make things work correctly or have to use some other synchronisation primitives?

mojoX911 commented on src/maker/api.rs:
```diff
@@ -92,6 +91,48 @@ pub struct ConnectionState {
+            .lock()
+            .map_err(|_| MakerError::General("Failed to lock threads"))?;
+        while let Some(thread) = threads.pop() {
+            thread.join().unwrap();
```
Instead of unwrap, how about just `log::error` the error. The error later is being suppressed anyway in the drop implementation, so we can as well just log it here. 

Currently, this loop will return at the first instance of the error and will not try to join the rest of the threads. So lets just log the error here if any, or else just say "Thread X terminated successfully".

claddyy replied:
Yep, makes sense.

mojoX911 reviewed (commented):
I like the structure. Ack. 

Below is one non-blocking comment. Can be done in a followup PR too, but will be good to have to help for debugging later.

mojoX911 commented:
> Also there are many instances in codebase where we are not handling the thread's error correctly -> so we incorporate those enhancements in this pr itself? For eg:
> 
> https://github.com/citadel-tech/coinswap/blob/6e5ce00f710685cd43b44a78896b7d1a3c6c6648/tests/test_framework/mod.rs#L146-L149
> 
> Here unwraping the `start_directory_server` inside the thread doesn't make sense -> as we fail to start directory server but still we continue with further procedure like initing a maker & taker etc and that's not good. Instead , we must propagate the error to main thread and unwrap there -> this will panic the test at that stage itself which will be good UX for tests.

Not needed for the test framework. The only other place where it makes sense is the DNS server when it creates the rpc thread. But not a major need as DNS is not user facing.

mojoX911 commented:
Merging this to unblock downstream works.

claddyy commented:
> I like the structure. Ack.
> 
> Below is one non-blocking comment. Can be done in a followup PR too, but will be good to have to help for debugging later.

Addressed the review here https://github.com/citadel-tech/coinswap/pull/324.
